### PR TITLE
add compression when posting metrics to dd

### PR DESCRIPTION
**Pull request by robpickerill**
Adds gzip compression when posting metrics to Datadog.

### CODE REVIEW

**Comment by robpickerill** [comment]
I think adding a configuration option for compression maybe the best way to go here, will rework the PR to support that.

File: surfacers/datadog/proto/config.proto [hunk 1]
```diff
@@ -17,4 +17,7 @@ message SurfacerConf {
 
   // Datadog server, default: "api.datadoghq.com"
   optional string server = 4;
+
+  // Compress metrics before sending them to Datadog.
+  optional bool compress = 7 [default = true];
```
Comment (robpickerill) [inline surfacers/datadog/proto/config.proto:22]: Pending #357

File: surfacers/datadog/client.go [hunk 2]
```diff
@@ -119,3 +130,19 @@ func (c *ddClient) submitMetrics(ctx context.Context, series []ddSeries) error {
 
 	return nil
 }
+
+func compressPayload(b []byte) (*bytes.Buffer, error) {
+	var buf bytes.Buffer
+	gz, err := gzip.NewWriterLevel(&buf, gzip.BestCompression)
```
Comment (manugarg) [inline surfacers/datadog/client.go:136]: Can we use this function:
https://github.com/cloudprober/cloudprober/blob/9ecd8a25aeb66c289d8b21f052d36ab89fbc9bf2/surfacers/common/compress/compress.go#L99

One diff is base64 conversion, in the Compress function, gzipped content is converted to base64. I think we don't need that here? If that's the case let's keep it as it is.

**Review by manugarg — COMMENTED** [review]
Looks good. Just one minor suggestion.

File: surfacers/datadog/client.go (hunk 2 above)
Comment (robpickerill) [reply surfacers/datadog/client.go:136]: Yeah I see, I didn't spot that function and it's a little duplication here. 

But I don't think the base64 encoding is required, but before we commit, let me try it with Datadog and come back to you.

Comment (robpickerill) [reply surfacers/datadog/client.go:144]: The base64 encoding was a problem when testing with Datadog, and I couldn't see a way to inform dd that the encoding type was base64 encoded and compressed. I opt'ed for the private compressPayload function in the end if your ok with that?

File: surfacers/datadog/proto/config.proto (hunk 1 above)
Comment (robpickerill) [reply surfacers/datadog/proto/config.proto]: I rewrote the config to drop the overwritten default value, and I refactored the PR a little to remove the double negative evaluations in the logic.

**Comment by robpickerill** [comment]
I was also thinking about adding a minimum size for compression, like 1kb or 2kb, let me know if you think that would be preferable here.

File: surfacers/datadog/client.go (hunk 2 above)
Comment (manugarg) [reply surfacers/datadog/client.go:144]: Yes. Thanks for checking!

**Review by manugarg — APPROVED** [review]
Thanks, Rob!

**Pull request merged**